### ISSUE TITLE
Having Trouble Compiling a module .dll

### ISSUE THREAD

**Issue by boagz57** [godotengine/godot#11766]
After cloning the master branch of godot yesterday (9/30/17) and building the engine for windows I tried going through the module tutorial. I was able to statically build and use the 'sumator' module (as in the tuorial) without problems, however, when trying to build a dll version of the module through VS2015, I get these error messages:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1181	cannot open input file 'sumator.windows.tools.32.windows.tools.32.lib'	godot	F:\godot\LINK	1	

Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command "echo Starting SCons && cmd /V /C set "plat=x86" ^& (if "x86"=="x64" (set "plat=x86_amd64")) ^& set "tools=yes" ^& (if "debug"=="release" (set "tools=no")) ^& call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" !plat! ^& scons platform=windows target=debug tools=!tools! -j1" exited with code 2.	godot	C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.MakeFile.Targets	41	
```
The only file that changed between my successful build of static module and this one was the SCsub file which was setup like the tutorial:

```
Import('env')

sources = [
    "register_types.cpp",
    "sumator.cpp"
]

module_env = env.Clone()
module_env.Append(CXXFLAGS='-fPIC')  # Needed to compile shared library
module_env['LIBS'] = []

shared_lib = module_env.SharedLibrary(target='#bin/sumator', source=sources)

shared_lib_shim = shared_lib[0].name.rsplit('.', 1)[0]
env.Append(LIBS=[shared_lib_shim])
env.Append(LIBPATH=['#bin'])
```

I know some of the tutorials are becoming dated with version 3.0, so is there something wrong with the file above or is there something else I need to do to get a .dll to show up in the bin folder?

**Comment by akien-mga** [comment]
CC @touilleMan who wrote that part of the docs: https://github.com/godotengine/godot-docs/commit/927ef4576ccfc9a184728620fe16ae9a46944f50

It's kinda moot now that we can use GDNative for this, and note that this whole paragraph is for advanced users who want to build their module as a shared library; the sections before that in the tutorial should be enough to build a static module.

**Comment by touilleMan** [comment]
I have to admit I never check if the shared lib compilation of a module worked on windows.
And after the pain we had with @karroffel with windows about linking and how .dll are much different beasts than .so, I'm not that surprised to see it doesn't work.

Anyway, as @akien-mga said this is totally deprecated in favor of GDNative, and should be removed from godot 3.0 documentation in favor of GDNative doc

**Comment by boagz57** [comment]
According to this [doc page](https://godotengine.org/article/dlscript-here) the author explains that GDNative is not there to replace modules and that modules still have there uses as they can 'extend' engine functionality where GDNative cannot. Is this still the case? Or has GDNative functionality been extended further?

**Comment by akien-mga** [comment]
> According to this doc page the author explains that GDNative is not there to replace modules and that modules still have there uses as they can 'extend' engine functionality where GDNative cannot. Is this still the case? Or has GDNative functionality been extended further?

That's correct, and modules are still supported. But that SCsub you're using is not the standard way to build a module, it's a hack to build it as a shared library (instead of the default of linking it statically in the engine). The use case for shared libraries is best covered by GDNative, as it's typically for performance-critical game logic.

If you're learning to make modules, just ignore that last section about making it a shared library.

**Comment by hungrymonkey** [comment]
```
$scons p=osx use_llvm=yes
YASM is necessary for WebM SIMD optimizations.
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[ 88%] Linking Program        ==> bin/godot.osx.tools.64
ld: library not found for -lsummator.osx.tools.64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.osx.tools.64] Error 1
scons: building terminated because of errors.

```

i am just here to say that osx shared library is broken too

is here my test code

https://github.com/hungrymonkey/godot/commit/121d78a1e6e64e6828d3f05932a302d4c8287b7a

**Comment by logicor-cn** [comment]
https://github.com/godotengine/godot-docs/issues/2099

Same issue here.  Posted a little investigation in there.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.1 or any later release (e.g. [3.2.2-beta4](https://godotengine.org/article/dev-snapshot-godot-3-2-2-beta-4))?

**Comment by KoBeWi** [comment]
Closing due to no reply. Please comment if you can still reproduce in 3.2.3+.

**Comment by marchani** [comment]
@KoBeWi: I can still reproduce this bug in Windows using the `3.2.3-stable` tag.
